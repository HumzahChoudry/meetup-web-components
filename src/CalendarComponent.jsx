import React from 'react';
import cx from 'classnames';
import Flatpickr from 'flatpickr';

/**
 * @module CalendarComponent
 * inits flatpickr js date picker over a text input
*/
class CalendarComponent extends React.Component {

	constructor(props) {
		super(props);
		this.state = {
			value: this.props.value || ''
		};
		this.onChange = this.onChange.bind(this);
		this.onOpen = this.onOpen.bind(this);
		this.onClose = this.onClose.bind(this);
	}

	/**
	* @description init the js date flatpickr component
	*/
	componentDidMount() {
		const options = {
			onChange: this.onChange,
			onOpen: this.onOpen,
			onClose: this.onClose,
			altInput: true,
			altFormat: 'D M d, Y', // TODO localize
			defaultDate: this.props.value
		};

		Object.assign(options, this.props.opts);
		this.flatpickr = new Flatpickr(this.node, options);
	}

	componentWillUnmount() {
		this.flatpickr && this.flatpickr.destroy();
	}

	/**
	* @function onChange
	* @param Array selectedDates
	* @param dateStr
	* @param Object instance the calendar instance
	* @description signature conforms to the onChange handler flatpickr expects
	* calls the callback with the selectedDates value (callback used in wrapping components)
	*/
	onChange(selectedDates, dateStr, instance) {
		this.setState({ value: selectedDates[0] });
		this.props.callback && this.props.callback(selectedDates[0]);
	}

	/**
	* @function onOpen
	* @description event hook for flatpickr, used to call onFocus
	* and apply focus highlight if this is a DateTimePicker
	*/
	onOpen() {
		this.props.onFocus && this.props.onFocus();
	}

	/**
	* @function onClose
	* @description event hook for flatpickr, used to call onBlur
	* and remove focus highlight if this is a DateTimePicker
	*/
	onClose() {
		this.props.onBlur && this.props.onBlur();
	}

	render() {
		const {
			callback,	// eslint-disable-line no-unused-vars
			className,
			id,
			name,
			value,		// eslint-disable-line no-unused-vars
			opts,		// eslint-disable-line no-unused-vars
			...other
		} = this.props;

		const classNames = cx(
			'input--dateTimePicker',
			className
		);

		return (
			<input
				id={id}
				type='text'
				name={name}
				defaultValue={this.state.value}
				className={classNames}
				ref={ node => this.node = node }
				{...other} />
		);
	}
}

CalendarComponent.propTypes = {
	name: React.PropTypes.string.isRequired,
	callback: React.PropTypes.func
};

export default CalendarComponent;

