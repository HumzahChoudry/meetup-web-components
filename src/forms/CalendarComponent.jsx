/* eslint-disable no-return-assign */
import PropTypes from 'prop-types';
import React from 'react';
import cx from 'classnames';
import withErrorList from '../utils/components/withErrorList';

/**
 * @module CalendarComponent
 * inits flatpickr js date picker over a text input
*/
export class CalendarComponent extends React.Component {
	constructor(props) {
		super(props);

		this.onOpen = this.onOpen.bind(this);
		this.onClose = this.onClose.bind(this);

		this.onFlatpickrChange = this.onFlatpickrChange.bind(this);
	}

	/**
	* @description init the js date flatpickr component
	*/
	componentDidMount() {
		// flatpickr uses `window` on import,
		// which breaks on server-sider render.
		// lazy-loading flatpickr ensures it is
		// imported only in clientside envs.
		const Flatpickr = require('flatpickr');
		const options = {
			onChange: this.onFlatpickrChange,
			onOpen: this.onOpen,
			onClose: this.onClose,
			altInput: true,
			allowInput: true,
			altFormat: 'D M d, Y',
			defaultDate: this.props.value,
			nextArrow: `<span class="svg svg--chevron-right">
				<svg preserveAspectRatio="xMinYMin meet" width="12" height="12" viewBox="0 0 12 12" className="svg-icon valign--middle" role="img">
				<use xlink:href="#icon-chevron-right" />
				</svg>
			</span>`,
			prevArrow: `<span class="svg svg--chevron-left">
				<svg preserveAspectRatio="xMinYMin meet" width="12" height="12" viewBox="0 0 12 12" className="svg-icon valign--middle" role="img">
				<use xlink:href="#icon-chevron-left" />
				</svg>
			</span>`,
			...this.props.datepickerOptions,
		};

		this.flatpickr = new Flatpickr(this.inputEl, options);
	}

	componentWillUnmount() {
		this.flatpickr && this.flatpickr.destroy();
	}

	// replaces updateFlatpickr
	// if we receive a new value from parent, update Flatpickr
	componentWillReceiveProps(newProps) {
		if (this.flatpickr) {
			if (newProps.datepickerOptions) {
				this.updateFlatpickrOptions(newProps.datepickerOptions);
			}
			this.flatpickr.setDate(newProps.value);
		}
	}

	/**
	 * Updates config options on the flatpickr instance
	 * @param {Object} newOptions new configaration options for flatpickr
	 * @return {undefined}
	 */
	updateFlatpickrOptions(newOptions) {
		Object.keys(newOptions)
			.filter(
				option => this.props.datepickerOptions[option] !== newOptions[option]
			)
			.forEach(option => {
				this.flatpickr.set(option, newOptions[option]);
			});
	}

	/**
	* @function onFlatpickrChange
	* @param Array selectedDates
	* @param dateStr
	* @param Object instance the calendar instance
	* @description signature conforms to the onChange handler flatpickr expects
	* calls onChange if prop provided (eg from redux-form or wrapping components like DateTimePicker)
	*/
	onFlatpickrChange(selectedDates, dateStr, instance) {
		this.props.onChange && this.props.onChange(selectedDates[0]);
	}

	/**
	* @function onOpen
	* @description event hook for flatpickr, used to call onFocus
	* and apply focus highlight if this is a DateTimePicker
	*/
	onOpen() {
		this.props.onFocus && this.props.onFocus();
	}

	/**
	* @function onClose
	* @description event hook for flatpickr, used to call onBlur
	* and remove focus highlight if this is a DateTimePicker
	*/
	onClose() {
		this.props.onBlur && this.props.onBlur();
	}

	render() {
		const {
			className,
			name,
			label,
			required,
			value,
			datepickerOptions, // eslint-disable-line no-unused-vars
			onChange, // eslint-disable-line no-unused-vars
			helperText,
			...other
		} = this.props;

		const classNames = {
			label: cx(
				{
					required,
					'flush--bottom': helperText
				},
				className
			),
			helperText: cx(
				'helperTextContainer',
				{ required }
			),
			field: cx(
				'input--dateTimePicker select--reset',
				className
			)
		};


		return (
			<span>
				{label && (
					<label htmlFor={name} className={classNames.label}>
						{label}
					</label>
				)}
				{helperText &&
					<div className={classNames.helperText}>
						{helperText}
					</div>
				}
				<input
					type="text"
					id={name}
					name={name}
					defaultValue={value}
					className={classNames.field}
					ref={input => (this.inputEl = input)}
					{...other}
				/>
			</span>
		);
	}
}

CalendarComponent.propTypes = {
	name: PropTypes.string.isRequired,
	onChange: PropTypes.func, // provided by DateTimePicker or redux-form
	datepickerOptions: PropTypes.object,
	suppressError: PropTypes.bool,
	helperText: PropTypes.oneOfType([
		PropTypes.string,
		PropTypes.element
	])
};

CalendarComponent.defaultProps = {
	datepickerOptions: {},
	suppressError: false,
};

export default withErrorList(CalendarComponent);
