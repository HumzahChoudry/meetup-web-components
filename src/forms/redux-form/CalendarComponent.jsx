import React from 'react';
import CalendarComponent from '../CalendarComponent';

/**
 * @param {Object} props - props passed in from parent and redux-form
 * @description wraps standard TextInput web component for use with redux-form
 * deconstructs props that redux-forms sets and sets them on TextInput
 * @return {Component} TextInput
 */
const ReduxFormCalendarComponent = props => {
	const { meta, input, validateBeforeTouched, ...other } = props;
	const error = (validateBeforeTouched || meta.touched) ? meta.error : null;

	return <CalendarComponent error={error} {...input} {...other} />;
};

ReduxFormCalendarComponent.displayName = 'ReduxFormCalendarComponent';

export default ReduxFormCalendarComponent;
