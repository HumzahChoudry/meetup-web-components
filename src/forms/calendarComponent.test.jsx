import React from 'react';
import { shallow } from 'enzyme';
import { CalendarComponent } from './CalendarComponent';

describe('CalendarComponent', function() {
	let calendarComponent;

	const id = 'beyonce',
		name = 'halo',
		className = 'beyonce-halo',
		value = new Date('2012-10-12'),
		newValue = new Date('2020-10-12'),
		opts = { dateFormat: 'm-d-Y' };

	const onChangeCallback = jest.fn();

	it('renders the calendarComponent with appropriate attributes', function() {
		calendarComponent = shallow(
			<CalendarComponent
				id={id}
				name={name}
				className={className}
				value={value}
				datepickerOptions={opts}
				onChangeCallback={onChangeCallback}
				error="this is an error"
				ref={node => (this.node = node)}
			/>
		);
		calendarComponent.instance().componentDidMount();
		expect(calendarComponent).toMatchSnapshot();
		calendarComponent = null;
	});

	describe('onChange callback tests', () => {
		let calendarComponent, onChangePropMock, callbackMock;

		// because we are rendering shallow here, flatpickr isnt
		// instantiated (because componentDidMount not called)
		// just using direct calls to flatpickr change to test

		it('calls onChange prop if one is provided, as with redux-form', function() {
			onChangePropMock = jest.fn();
			callbackMock = jest.fn();

			calendarComponent = shallow(
				<CalendarComponent
					id={id}
					name={name}
					className={className}
					onChange={onChangePropMock}
					onChangeCallback={callbackMock}
				/>
			);
			expect(onChangePropMock).not.toHaveBeenCalled();
			calendarComponent.instance().onFlatpickrChange([newValue]);
			expect(onChangePropMock).toHaveBeenCalledWith(newValue);
			calendarComponent = null;
		});
	});

	describe('onBlur onFocus callback tests', () => {
		let cal, onFocusMock, onBlurMock;

		beforeEach(() => {
			onFocusMock = jest.fn();
			onBlurMock = jest.fn();

			cal = shallow(
				<CalendarComponent
					id={id}
					name={name}
					className={className}
					onFocus={onFocusMock}
					onBlur={onBlurMock}
				/>
			);
		});

		afterEach(() => {
			cal = null;
			onFocusMock.mockClear();
			onBlurMock.mockClear();
		});

		it('onOpen calls the props onFocus method when it exists', () => {
			expect(onFocusMock).not.toHaveBeenCalled();
			cal.instance().onOpen();
			expect(onFocusMock).toHaveBeenCalled();
		});

		it('onClose calls the props onBlur method when it exists', () => {
			expect(onBlurMock).not.toHaveBeenCalled();
			cal.instance().onClose();
			expect(onBlurMock).toHaveBeenCalled();
		});
	});

	describe('componentWillReceiveProps', () => {
		it('should updateFlatpickrOptions when receiving datepickerOptions prop', () => {
			const instance = shallow(<CalendarComponent name="test" />).instance();
			const datepickerOptions = { option1: true };

			instance.flatpickr = { setDate: jest.fn() };

			jest
				.spyOn(instance, 'updateFlatpickrOptions')
				.mockImplementation(() => {});

			instance.componentWillReceiveProps({ datepickerOptions });

			expect(instance.updateFlatpickrOptions).toHaveBeenCalledWith(
				datepickerOptions
			);
		});
	});

	describe('updateFlatpickrOptions', () => {
		it('should update flatpickr with updated options', () => {
			const optionThatDoesntChange = 'will not change';

			const wrapper = shallow(
				<CalendarComponent
					name="test"
					datepickerOptions={{
						optionThatDoesntChange,
						maxDate: 1507751589972,
					}}
				/>
			);

			const instance = wrapper.instance();

			instance.flatpickr = {
				set: jest.fn(),
			};

			const minDate = 12345;
			const maxDate = 67890;

			instance.updateFlatpickrOptions({
				minDate,
				maxDate,
				optionThatDoesntChange,
			});

			expect(instance.flatpickr.set).toHaveBeenCalledWith('maxDate', maxDate);
			expect(instance.flatpickr.set).toHaveBeenCalledWith('minDate', minDate);
			expect(instance.flatpickr.set).not.toHaveBeenCalledWith(
				'optionThatDoesntChange',
				optionThatDoesntChange
			);
		});
	});
});
