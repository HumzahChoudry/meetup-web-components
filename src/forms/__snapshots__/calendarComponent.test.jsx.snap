// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CalendarComponent renders the calendarComponent with appropriate attributes 1`] = `
<div>
  <span>
    <input
      className="input--dateTimePicker select--reset beyonce-halo"
      defaultValue={2012-10-12T00:00:00.000Z}
      id="beyonce"
      name="halo"
      onChangeCallback={[Function]}
      type="text"
    />
  </span>
</div>
`;
